### ISSUE TITLE
a11y tree

### ISSUE THREAD

**Issue by mfulton26** [dequelabs/axe-core#882]
I'm looking for a way to get something similar to the a11y tree in Chrome DevTools's [a11y pane](https://developers.google.com/web/updates/2018/01/devtools#a11y-pane) except I want to compute the tree via JavaScript in Chrome, Edge, Firefox, Safari, and, if possible, IE11.

From what I can tell aXe does not create a formal a11y tree but creates a flattened tree for its Virtual DOM and then performs its audits using it (please correct me if I've missed something).

Is there a way to have aXe dump a formal a11y tree with a11y nodes (which may map to zero or more actual DOM nodes) along with their roles, aria properties, etc.?

Ultimately I am looking for a cross-browser, JavaScript a11y tree util to be used for primarily two reasons:
1. To programmatically traverse the tree to locate widgets preceded by text for more human-like location strategies to be used via Selenium WebDriver (e.g. `webDriver.findElement(context({label: "First Name"}))` instead of `webDriver.findElement({id: "fname"})`). I have something like this working now (see https://github.com/mfulton26/webautomator) but it does not have full support for aria markup, etc. If I had a JavaScript-written screen reader (so to speak) then I could use it instead to scan and find elements I want to interact with.
2. To dump a representation of the page to use in [jest-snapshot](https://facebook.github.io/jest/docs/en/snapshot-testing.html) testing. Using the actual DOM has too much variable markup (e.g. as development on a page continues different class names are added/removed, varying attributes are defined for script callbacks, etc. all while an a11y tree would be mostly untouched unlike the actual DOM).

If there are other tools anyone knows of that might fit my needs in addition to (or instead of) aXe that would be very helpful to know as well.

Any help on this would be greatly appreciated.

**Comment by WilcoFiers** [comment]
Your analyses is spot on. Axe-core creates a flattened DOM tree so that it can test inside of open shadow trees. It doesn't generate the accessibility tree, so far I don't think we've needed it. At most what I can see us do is that we add methods for grabbing owned / owner elements. That shouldn't be too difficult to write. Axe-core's `getAccessibleText` method also isn't the same as the accessible name computation algorithm, and we don't have a method to compute the role. I actually would argue that you can't compute the role, at best you could guess it based on known support, but its pretty flakey.

I would suggest you take axe-core's flattened tree write your own `getOwnerElement()` / `getOwnedElements()` methods to go with it. I don't know of any other tools that generate the flattened tree, which you'll need if you want to work with shadow DOM. If you're interested in writing something like this, I'd be like to work with you to see how we could put this inside axe-core. It sounds like a useful addition to the existing APIs.

**Comment by WilcoFiers** [comment]
Oh, and also, you may be interested in this project: https://github.com/xi/accname-test/ its a test suite for the accessible name computation algorithm.

**Comment by WilcoFiers** [comment]
Closing this ticket due to inactivity.
